### ISSUE TITLE
DDEV-Local v1.16 Test Plan

### ISSUE THREAD

**Issue by rfay** [drud/ddev#2599]
It's testing time again here in the DDEV-Local factory.  We'd love community help with every item that is interesting or relevant to you.  Please use [ddev v1.16.0-rc2](https://github.com/drud/ddev/releases/tag/v1.16.0-rc2) for your testing, it's available in homebrew (macOS, Linux, WSL2) and Chocolatey, and of course you can install other ways as shown in the release notes.

If you find a bug that you're pretty sure is a bug, then just open an issue for it.

To give feedback or ask questions, please just reply on this issue or discuss in slack or another other support channel.

1. `webroot` and `html` directories candidates for automatic docroot detection [#2434](http://github.com/drud/ddev/issues/2434) - In an empty project:
    - `mkdir webroot` or html and place in index.php in it
    - `ddev config` should discover it as the default docroot
2. Allow processing of PDF files with ImageMagick ([#2431](https://github.com/drud/ddev/pull/2431))
    - In a TYPO3 project back-end, visit Environment→Image Processing→Test Images.  All the images should show OK, but "Read pdf" is the one that is fixed by this issue.
3. Speed up ddev commands including ddev list, especially on macOS, fixes #1881 ([#2445](https://github.com/drud/ddev/pull/2445))
    - With a few projects going, `ddev list` on v1.15.3 and on current. Compare times.
    - Review other ddev command differences
4. Allow multiple directives including OSType and ProjectType in custom command, fixes [#2209](https://github.com/drud/ddev/issues/2209) ([#2452](https://github.com/drud/ddev/pull/2452))
    - Review custom commands that use OSType and ProjectType; there are several provided in ~/.ddev. They should show up only on the assigned OSType or ProjectType.
5. Use docker api to push CA keys to ddev-router, general ddev-router maintenance, fixes [#2398](https://github.com/drud/ddev/issues/2398) ([#2456](https://github.com/drud/ddev/pull/2456))
6. Verify that nginx in ddev-webserver and ddev-router are v1.18.0 (`nginx -V`)
7. Verify that ddev-router and ddev-webserver accept TLS1, 1.1, 1.2, 1.3  `nmap --script ssl-enum-ciphers -p 443 [junk.ddev.site](http://junk.ddev.site)` and after `brew upgrade openssl` and `export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"` `openssl s_client -connect junk.ddev.site:443 -tls1_3` (or tls1 or tls1_1 or tls1_2)
8. typo3 and typo3cms commands:
    - `typo3` and `typo3cms` commands should only be listed on projects of type TYPO3 (but they should be listed there)
    - `ddev typo3` and `ddev typo3cms` commands should give expected results
9. Custom TLS certificate management ([#2458](https://github.com/drud/ddev/pull/2458))
    - Add a custom cert to a project, probably a self-signed certificate or from another source
    - Verify that the custom cert is used by the project
    - Verify that the custom cert is removed when it should be
8. Let's Encrypt experimental support ([#2458](https://github.com/drud/ddev/pull/2458)) (see [docs](https://ddev.readthedocs.io/en/latest/users/alternate-uses/#project-hosting-on-the-internet-including-lets-encrypt))
    - Put a project on a server
    - Enable casual webhosting for the project using [docs](https://ddev.readthedocs.io/en/latest/users/alternate-uses/#project-hosting-on-the-internet-including-lets-encrypt), enable Let's Encrypt
    - Verify expected secure behavior; inspect cert.
9. Verify that *.rss + *.atom are handled #2403 ([#2461](https://github.com/drud/ddev/pull/2461))
10. Verify headers to see that css/js are not cached by browser ([#2413](https://github.com/drud/ddev/pull/2413))
11. Verify that wildcard additional_hostnames, fixes #2391 ([#2393](https://github.com/drud/ddev/pull/2393)). 
    - Make sure the feature works
    - Check for docs
12. Check for standard Drupal/Backdrop  settings.php based on defaults, fixes [#1097](https://github.com/drud/ddev/issues/1097) ([#2500](https://github.com/drud/ddev/pull/2500))
    - New drupal* and backdrop projects starting with no settings.php should get a settings.php based on defaults, not an abbreviated one.
13. Verify that DRUSH_OPTIONS_URI is not set if disable_settings_management=true, fixes [#2443](https://github.com/drud/ddev/issues/2443) ([#2506](https://github.com/drud/ddev/pull/2506))
14. ddev-router should not accept hostnames it's not configured for, fixes [#2345](https://github.com/drud/ddev/issues/2345) ([#2504](https://github.com/drud/ddev/pull/2504))
    - Start two projects, stop one. Hit the PHPMyAdmin URL for the stopped one. You shouldn't get anything.
15. Verify docker-compose overrides for router-compose.yaml and ssh-auth-compose.yaml**, fixes [#1779](https://github.com/drud/ddev/issues/1779), fixes** [#2476](https://github.com/drud/ddev/issues/2476) ([#2505](https://github.com/drud/ddev/pull/2505))
    - Use an override for router-compose.anotherimage.yaml or something. Verify that the correct router image is used
    - Same thing for ssh-auth-compose.yaml
16. Add the option to add flags to custom commands, fixes #2491 (#2493)
    - Add a flag to a custom command
    - Verify behavior
17. ARM64 installation and behavior
    - Install on arm64 machine using install_ddev.sh
    - Verify correct behavior
18. `ddev tableplus` command  #2522:  (#2526) (macOS only)
19. Images of omitted containers should not be pulled, fixes #2262 (#2541)
20. Drupal 8 projects should use MariaDB 10.3 by default ([#2372](https://github.com/drud/ddev/pull/2372))
21. Experimental PHP8.0 support, fixes #2553 (#2564)
    - Configure a simple project for php8.0, verify behavior (Drupal 9.1 works OK with PHP8)
    - Verify php8.0-rc3 (or perhaps higher)
    - Use xdebug; you should be able to step-debug as always
22. Shopware6 support for ddev, make hostnames and URLs lowercase, fixes [#1988](https://github.com/drud/ddev/issues/1988) ([#2551](https://github.com/drud/ddev/pull/2551))
    - Create a Shopware6 project using [docs](https://ddev.readthedocs.io/en/latest/users/cli-usage/#shopware-6-quickstart)
    - Click around and see how it goes.
23. ddev auth ssh shouldn't work if ddev-ssh-agent opted out, fixes [#2573](https://github.com/drud/ddev/issues/2573) ([#2576](https://github.com/drud/ddev/pull/2576))
    - `omit_containers: [ddev-ssh-agent]`
    - `ddev auth ssh` should say it can't do it.
24. Fix missing output of exec-host and make it interactive ([#2579](https://github.com/drud/ddev/pull/2579))
    - Create an exec-host action on a post-start hook; it should have some output
    - Verify that you see the output on start
    - Edit the action to wait for input and verify that the input is accepted and start continues.
25. composer version setting:
    - Verify default composer version should be 1.x `ddev exec composer --version`
    - Set to v2 with `ddev config --composer-version=2` and `ddev start` and `ddev exec composer --version`
    - Set back to v1 with `ddev config --composer-version=1` and `ddev start` and `ddev exec composer --version`
    - Set to an explicit version with `ddev config --composer-version=1.10.15`, start, verify `ddev exec composer --version`
    - Do a self-update with `ddev exec composer self-update` and verify results.
26. Hardened images, casual webhosting for ddev, fixes #2446 (#2589)
    - Use `ddev config global --use-hardened-images` and `ddev start` and `ddev exec sudo ls` should fail.
    - Review security situation on server set up in #8
27. ddev-ssh-agent and `ddev auth ssh` improvements, fixes [#2597](https://github.com/drud/ddev/issues/2597), fixes [#2608](https://github.com/drud/ddev/issues/2608) ([#2606](https://github.com/drud/ddev/pull/2606))
    - `ddev poweroff` and then `ddev auth ssh` should work OK
    - If there are several keys in hosts ~/.ssh, all should be loaded on `ddev auth ssh`
28. Magento2 Quickstart improvements, fixes [#2600](https://github.com/drud/ddev/issues/2600), fixes [#2602](https://github.com/drud/ddev/issues/2602) ([#2603](https://github.com/drud/ddev/pull/2603))
    - Review [Magento 2 quickstart](https://ddev.readthedocs.io/en/latest/users/cli-usage/#magento-2-quickstart) and run through it.
29. `ddev composer create psr/log -y` and `ddev delete images -y` should work, Add --yes -y flags to composer create and delete images, fixes [#2387](https://github.com/drud/ddev/issues/2387) ([#2609](https://github.com/drud/ddev/pull/2609))
30. Performance comparison: Make sure `ddev start` and webserving benchmarks remain the same as v1.15.3

**Comment by rfay** [comment]
My first pass:

1. ✅ **webroot and html directory docroot detection**: Worked fine.
2. ✅ **Allow processing of PDF files** with ImageMagick: PDF example shows fine in TYPO3 v10 Environment→Image Processing section.
3. ✅ **Command speedup on macOS**: 
    - 5 projects running; `ddev list` v1.16.0-rc2: 0.6s. v1.15.3: 5s
    - 5 projects (poweroff): `ddev list` v1.16.0-rc2: 0.3s   v1.15.3: 1.1s
    - Nothing running  (poweroff), but images pre-pulled `ddev start` v1.16.0-rc2: 17.9-22s  v1.15.3: 16.3-19s

4. ✅ **OSType, ProjectType, HostBinaryExists in commands**: 
    - macOS, SequelPro, SequelAce, TablePlus installed: All three show in generic project.  drush and typo3 commands do not show up.
    - wsl2 TYPO3 project: None of the sequel commands show, typo3 and typo3cms commands show up.
    - wsl2 Drupal project: None of sequel commands show; typo3 commands do not show; drush does show.
5. ✅ **nginx versions**: Both are 1.18.0
6. ✅Use docker api to push CA keys to ddev-router: Working
7. 👎Verify that ddev-router and ddev-webserver accept TLSv1.3 but does not accept TLS1, 1.1.  `nmap --script ssl-enum-ciphers -p 443 [junk.ddev.site](http://junk.ddev.site)` and after `brew upgrade openssl` and `export PATH="/usr/local/opt/openssl@1.1/bin:$PATH"` `openssl s_client -connect junk.ddev.site:443 -tls1_3` (or tls1 or tls1_1 or tls1_2)
    - Despite the current settings, both accept TLSv1 and TLSv1.1. I believe this has something to so with SNI, and that the ssl_protocols directive should be placed in the http context (in nginx.conf) rather than the server context.
    - This really isn't very important for regular dev work, but is important for casual webhosting.
    - Edit: I think this might be a mistake doing this anyway. I made the change to nginx.conf on a production server and ended up with Chrome not being able to access the site.
8. ✅ typo3 and typo3cms commands only show up on typo3 project
    - `ddev typo3 site:list` works
    - `ddev typo3cms` only works after you `ddev composer require helhum/typo3-console`
    - `ddev typo3cms site:list` works
9. 👎 Custom TLS certificate management: 
    - I'm unable to get this to work.  I verify that the certs in the ddev-router nginx configuration are the right ones, but nginx is not serving based on those certs (it's using master.crt)
10. ✅ Casual webhosting with Let's Encrypt experimental support: Worked well, but I did have to `ddev poweroff` and `ddev start` a time or two. It's important to get DNS and firewall set up earlier than I did so Let's Encrypt can do what it needs to. Added a PR with small improvements to docs, emphasizing again that this is not professional managed hosting.
11. ✅ Verify that *.rss + *.atom are handled properly #2403 ([#2461](https://github.com/drud/ddev/pull/2461)). This gets a TYPO3 404 "The page did not exist or was inaccessible. Reason: The requested page does not exist"
12. ✅Verify headers to see that css/js are not cached by browser ([#2413](https://github.com/drud/ddev/pull/2413))
    - On a Drupal 8 Umami site I see card.css having response header "cache-control: no-cache", and it gets downloaded again on every visit to the site.
    - I see drupal.js get loaded on every visit, also "cache-control: no-cache".
    - However, I see both of these showing as a 304. But a minor change to drupal.js results in next page load downloading it as a 200.
13. ✅Wildcard additional hostnames fixes #2391 ([#2393](https://github.com/drud/ddev/pull/2393)).  This configuration results in [a.d88.ddev.site](http://a.d88.ddev.site), b.d88.ddev.site, a.junker99.ddev.site, [z.junker99.ddev.site](http://z.junker99.ddev.site) working. The docs are there in the [additional_hostnames section](https://ddev.readthedocs.io/en/latest/users/extend/additional-hostnames/#additional-project-hostnames).

    ```yaml
    additional_hostnames:
    - '*.d88'
    - '*.junker99'
    ```

14. ✅Check for standard Drupal/Backdrop  settings.php based on defaults, fixes [#1097](https://github.com/drud/ddev/issues/1097) ([#2500](https://github.com/drud/ddev/pull/2500))
    - I checked Drupal8 and Backdrop and they both had created settings.php files based on the original. I think it's good
15. ✅ Verify that DRUSH_OPTIONS_URI is not set if disable_settings_management=true, fixes [#2443](https://github.com/drud/ddev/issues/2443) ([#2506](https://github.com/drud/ddev/pull/2506)) (It's not set)
16. ✅ ddev-router should not accept hostnames it's not configured for, fixes [#2345](https://github.com/drud/ddev/issues/2345) ([#2504](https://github.com/drud/ddev/pull/2504))
    - I get a "503: No ddev back-end site available" after `ddev stop` on a project and trying to hit port 8037 for phpmyadmin.
17. ✅ Verify docker-compose overrides for router-compose.yaml and ssh-auth-compose.yaml**, fixes [#1779](https://github.com/drud/ddev/issues/1779), fixes** [#2476](https://github.com/drud/ddev/issues/2476) ([#2505](https://github.com/drud/ddev/pull/2505))
    - I created simple router-compose.x.yaml file which changed the healthcheck test to be `["CMD", "true"]`, short-circuiting the healtcheck. It worked.
    - I created a simple ssh-auth-compose.x.yaml that changed the username in the build args to "rufus" and the name of the user inside the container was "rufus".
18. ❓ Add the option to add flags to custom commands, fixes #2491 (#2493) (So far I haven't figured out how you implement behavior based on flags looking at [https://github.com/drud/ddev/pull/2493](https://github.com/drud/ddev/pull/2493) and [https://ddev.readthedocs.io/en/latest/users/extend/custom-commands/#flags](https://ddev.readthedocs.io/en/latest/users/extend/custom-commands/#flags)
19. ✅ ARM64 installation and behavior
    - Installed successfully with `./install_ddev.sh v1.16.0-rc2`
    - Did a fresh D7 site, tested out installation, PHPMyAdmin, MailHog
    - There are things people have to know:
        - How to install OS (for example on Raspberry Pi, need 64-bit)
        - docker-compose may need to be installed via pip3 (certainly does on Debian 10/Raspberry Pi 64-bit)
        - You have to install with the install script, as brew is not available.
20. ✅`ddev tableplus` command  #2522:  (#2526) (macOS only) (Command worked fine)
21. ✅ Images of omitted containers should not be pulled, fixes #2262 (#2541): I did `ddev config global --omit-containers=ddev-ssh-agent,dba` after deleting all images. Then `ddev start` and checked to see the result. The omitted images were not pulled.
22. ✅Drupal 8 projects should use MariaDB 10.3 by default ([#2372](https://github.com/drud/ddev/pull/2372)): I deleted the .ddev directory of an existing Drupal8 project and did `ddev config` and checked the config; it pulled and used 10.3. 
23. ✅Experimental PHP8.0 support, fixes #2553 (#2564): Ran with Drupal 9.1.x; had to composer install with php 7.4, but it ran ok with php 8.0
24. ✅📃 Shopware6 support for ddev, make hostnames and URLs lowercase, fixes [#1988](https://github.com/drud/ddev/issues/1988) ([#2551](https://github.com/drud/ddev/pull/2551)): This worked fine, except that I tried shopware 6.3, where the key difference is that the environment variable APP_URL must be set. 
25. ✅ddev auth ssh shouldn't work if ddev-ssh-agent opted out, fixes [#2573](https://github.com/drud/ddev/issues/2573) ([#2576](https://github.com/drud/ddev/pull/2576)). 
26. ✅Fix missing output of exec-host and make it interactive ([#2579](https://github.com/drud/ddev/pull/2579)): Added an exec-host on pre-start with "ls" and the output was shown. This worked fine, showed output, waited for input.

    ```yaml
    hooks:
      pre-start:
              - exec-host: 'ls && echo "please hit return" && read x && echo "YOU HIT IT"'

    ```

27. ✅composer version setting: Works great, used all along the way.
28. ✅Hardened images, casual webhosting for ddev, fixes #2446 (#2589): These are working, in use for hosting [randyfay.com](http://randyfay.com) and hobobiker.com
29. ✅ ddev-ssh-agent and `ddev auth ssh` improvements, fixes [#2597](https://github.com/drud/ddev/issues/2597), fixes [#2608](https://github.com/drud/ddev/issues/2608) ([#2606](https://github.com/drud/ddev/pull/2606))
    - `ddev poweroff && ddev auth ssh` worked; loaded all keys in ~/.ssh
    - `ddev auth ssh -d ~/tmp/junk` worked and loaded all keys
    - `ddev auth ssh` worked on Docker for Windows with Hyper-V mode enabled.
30. ✅`ddev composer create psr/log -y` and `ddev delete images -y` should work, Add --yes -y flags to composer create and delete images: Both worked great.

**Comment by rfay** [comment]
I think everything critical checks out, so cutting the release. There are a couple of followup issues, 

* https://github.com/drud/ddev/pull/2603#issuecomment-725756659 
* https://github.com/drud/ddev/issues/2633 
* https://github.com/drud/ddev/issues/2631
